### PR TITLE
The Trigger class and modifications to how commands are decorated and invoked

### PR DESCRIPTION
**Pull request by MHajoha**
### Changes:
* Add the server's port, whether to use tls or not and the channels to join upon connecting to the config.
* Implement the `Trigger` class to hold info on who invoked a command and where they did it. (+ a neat function for replying)
* Rework parts of command invokation for expandability in the future.
### Command invocation:
Commands now go through a chain of one or more decorators. The topmost (or only) decorator must be `Commands.command` to register the command. Each decorator' s wrapper function takes four arguments:
* `bot`: The bot receiving the command. Straightforward enough.
* `trigger`: `Trigger` object with info on the command sender etc.
* `words`: List of strings, the first of which is the name of the invoked command (with the prefix already stripped). The others are the arguments to the command, wich are seperated by whitespace.
* `words_eol`: List of strings much like `words`, but including everything from the argument to the end of the line.

The message `!mycommand arg1 arg2` would result in:
```
words = ["mycommand", "arg1", "arg2"]
words_eol = ["mycommand arg1 arg2", "arg1 arg2", "arg2"]
```
(if the command prefix is `!`, of course.)

Each decorator calls the underlying function with all four arguments (calling the wrapper created by the decorator below) or only `bot, trigger` if it is the bottommost decorator in the chain (calling the command function directly). 
An exception to this will likely be the parametrize decorator, which will, if present, always be at the bottom, calling the command function with `bot, trigger` followed by the parameters specified.

### CODE REVIEW

File: tests/test_permissions.py [hunk 1]
```diff
@@ -64,3 +72,29 @@ def test_permission_not_equal(self):
...
+
+    @async_test
+    async def test_restricted_command(self):
+        await Commands.trigger("!restricted", "some_recruit", "#somechannel")
+        self.assertIn({
+            "target": "#somechannel",
+            "message": permissions.OVERSEER.denied_message
+        }, self.bot.sent_messages)
+
+        await Commands.trigger("!restricted", "some_ov", "#somechannel")
```
Comment (theunkn0wn1) [inline tests/test_permissions.py]: It appears you are testing the same bit of code with different parameters, these should be marked as sub-tests

File: Modules/trigger.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        self.ident = ident
+        self.hostname = hostname
+        self.realname = realname
+        self.away = away
+        self.account = account
+        self.identified = identified
+
+    @classmethod
+    def from_bot_user(cls, bot: pydle.BasicClient, nickname: str, target: str):
+        """Creates a `Trigger` object from a user dictionary as used by pydle."""
```
Comment (theunkn0wn1) [inline Modules/trigger.py]: please place the parameters in the docstring. also, methinks a type hint could be nice here ;)

File: main.py [hunk 3]
```diff
@@ -86,7 +88,7 @@ class MechaClient(Client):
 
 
 @Commands.command("ping")
-async def cmd_ping(bot: Client, channel: str, sender: str):
+async def cmd_ping(bot, trigger):
     """
     Pongs a ping. lets see if the bots alive (command decorator testing)
     :param bot: Pydle instance
```
Comment (theunkn0wn1) [inline main.py]: Please update docstring, as you replaced parameters without touching the documentation,

**Review by theunkn0wn1 — CHANGES_REQUESTED** [review]
Looks good barring a few minor things. Can you please see to these?

File: Modules/trigger.py (hunk 2 above)
Comment (MHajoha) [reply Modules/trigger.py]: Are the names bot, nickname and target not self-explanatory? Anything more than a one-liner strikes me as a waste of space here.

Comment (UncleClapton) [reply Modules/trigger.py]: 23:09 <&xlexious> we are not programming on a fucking bbc micro with 4kb of memory

Comment (theunkn0wn1) [reply Modules/trigger.py]: By your logic Unit tests are a waste of space.
By your logic all documentation is a waste of space.

Comment (UncleClapton) [reply Modules/trigger.py]: 23:09 <&xlexious> we are not programming on a f*****g bbc micro with 4kb of memory

**Comment by theunkn0wn1** [comment]
oh come on, what did i break this time?

File: Modules/trigger.py (hunk 2 above)
Comment (MHajoha) [reply Modules/trigger.py]: No, by my 'logic' (I was asking a goddamn question.) redundant info in docstrings is a waste of space. And I'm not talking about memory >_<.

**Review by theunkn0wn1 — APPROVED** [review]

**Comment by theunkn0wn1** [comment]
LGTM.

**Pull request merged**